Add water fugacity to diffusion creep

danieldouglas92 opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

*Describe what you did in this PR and why you did it.*

### Before your first pull request:

* [ ] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [ ] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [ ] I have tested my new feature locally to ensure it is correct.
* [ ] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [ ] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -91,11 +105,34 @@ namespace aspect
+                                                  (constants::gas_constant*temperature)) *
+                                         std::pow(grain_size, p.grain_size_exponent);
+	  }
+
```
The `compute_strain_rate_and_derivative` function also needs to be modified.

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -222,8 +288,34 @@ namespace aspect
+										  expected_n_phases_per_composition);
         grain_size = prm.get_double("Grain size");
 
+	use_water_fugacity_diffusion = prm.get_bool("Use water fugacity for diffusion creep");
```
This change to the rheology needs to be documented somewhere.

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -91,11 +105,34 @@ namespace aspect
+
+	if (use_water_fugacity_diffusion)
+	  {
+	     AssertThrow(this->introspection().compositional_name_exists("water"),
```
I'm not sure "water" is a good name for the compositional field. Water is a free phase, not a chemical component in a solid. Perhaps "H2O concentration" would be more appropriate?

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -91,11 +105,34 @@ namespace aspect
+            double water_content = this->introspection().component_indices.compositional_fields[water_idx];
+
+            viscosity_diffusion = 0.5 / (p.prefactor / std::pow(p.wet_prefactor, p.water_fugacity_exponent)) *
+		                         std::pow(water_content, -p.water_fugacity_exponent) *
```
I'd suggest changing the sign of `p.water_fugacity_exponent`. It makes little sense to raise water concentration to a negative of a fugacity exponent.

danieldouglas92 replied:
Wouldn't this require inputting a negative water_fugacity_exponent in the parameter file though? It seems like mathematically it'll do the same thing but it'll be more confusing when inputting values for models

bobmyhill replied:
Mea culpa, I misread the equation. It's common to group the fugacity term with the prefactor (which is already part of the denominator of the first term), and if you did that, the negative sign would go away.

bobmyhill replied:
> I updated the code and addressed your comments @bobmyhill but at the moment the method John and I talked about required adding an argument to compute_viscosity that I hardcoded to be 1 inside rheology/composite_visco_plastic.cc and diffusion_dislocation.cc, definitely not ideal but thought it would be useful to push anyway to see if you had any thoughts on an approach

I agree that it isn't a beautiful solution, but I don't think it's a bad choice. The alternative would be to make a separate "wet diffusion" rheology, but that doesn't seem any better.

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -91,11 +105,34 @@ namespace aspect
+                         ExcMessage("The water fugacity rheology only works if "
+                                    "there is a compositional field called water."));
+            const unsigned int water_idx = this->introspection().compositional_index_for_name("water");
+            double water_content = this->introspection().component_indices.compositional_fields[water_idx];
```
```suggestion
            double water_concentration = this->introspection().component_indices.compositional_fields[water_idx];
```

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -176,6 +213,35 @@ namespace aspect
+                           "for a total of N+1 values, where N is the number of compositional fields. "
+                           "If only one value is given, then all use the same value. "
+                           "Units: \\si{\\meter\\cubed\\per\\mole}.");
+	prm.declare_entry ("Water fugacity exponents for diffusion creep", "1.0",
```
If you do change the sign of this exponent in the equations above, the default value should be -1, rather than 1.

bobmyhill reviewed (changes requested):
Looks pretty good to me.

A few minor comments. 

 A good place to document the rheology would be in the comment block in `compute_strain_rate_and_derivative` (the code in that block also needs updating). You should also provide a literature reference for the rheology.
 
An important note: this (common) formulation implies that H2O is a vital chemical component in allowing minerals to deform. Using this formulation, as H2O concentration -> 0, viscosity -> infinity. This is not a reasonable assertion. However, the formulation is empirically justified within a limited range of H2O concentrations. You should probably add a note to say that the `H2O concentration` compositional field should always have positive values to avoid infinite viscosities.

bobmyhill commented:
You will also need to:
- run the indent script after making the changes
- provide a new testcase and
- add a changelog entry.

naliboff commented:
> An important note: this (common) formulation implies that H2O is a vital chemical component in allowing minerals to deform. Using this formulation, as H2O concentration -> 0, viscosity -> infinity. This is not a reasonable assertion. However, the formulation is empirically justified within a limited range of H2O concentrations. You should probably add a note to say that the H2O concentration compositional field should always have positive values to avoid infinite viscosities.

danieldouglas92 commented:
I updated the code and addressed your comments @bobmyhill but at the moment the method John and I talked about required adding an argument to compute_viscosity that I hardcoded to be 1 inside rheology/composite_visco_plastic.cc and diffusion_dislocation.cc, definitely not ideal but thought it would be useful to push anyway to see if you had any thoughts on an approach

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -91,11 +106,41 @@ namespace aspect
+
+	if (use_water_fugacity_diffusion)
+	  {
+	     AssertThrow(this->introspection().compositional_name_exists("water_concentration"),
```
This assert should go into parse_parameters.

bobmyhill commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -91,11 +106,41 @@ namespace aspect
+	//    viscosity = 0.5 * A'^(-1) * d^(m) * C_H2O^(-r) * exp((E' + P*V')/(RT))
+	// A' = A / A_H2O^(r)
+	// E' = E - r*E_H2O
+	// V' = V - r*V_H2O
```
Now that I see this equation written down, most of the terms are pointless. A', E' and V' are all constants, so why not just require the user to put those in the prm file rather than make ASPECT calculate them billions of times?

I think the only term that needs to be in here is C_H2O^(-r), which is the standard term found in Mei and Kohlstedt (https://agupubs.onlinelibrary.wiley.com/doi/10.1029/2000JB900179), among others.

bobmyhill replied:
That would mean that you could just add a single new creep parameter (creep_parameters.water_fugacity_exponent), rather than four.

bobmyhill replied:
Sorry, in the cold light of day my comment from last night seems overly harsh. I could potentially see modifications to A, E and V being necessary, but perhaps via a relationship with `C_H2O`, rather than `r`? If you have a reference for this formulation, that would be great :)

danieldouglas92 replied:
No need to apologize I think you make a great point! The equations could only be modified to include the water concentration and if you just calculate the reduction in the creep parameters  before hand then only the water_fugacity_exponent would need to be defined and that would save a lot of lines of code. I don't have a formulation for modifying A, E or V with variable C_H2O but I'll see what I can dig up!

danieldouglas92 replied:
I didn't find a formulation for how to vary the creep parameters with varying C_H2O so I just simplified the existing code and included the same thing for dislocation creep, if you do know or come across something where A, E and V can be modified with C_H2O let me know! That would be very cool to try to implement

naliboff replied:
@bobmyhill - Are you ok with the current approach (user modifies values in PRM file), and then modify the flow law parameters automatically later on if we find a robust relationship with C_H20?

naliboff commented on source/material_model/diffusion_dislocation.cc:
```diff
@@ -84,8 +84,8 @@ namespace aspect
 
-              const std::pair<double, double> diff_edot_and_deriv = diffusion_creep.compute_strain_rate_and_derivative(stress_ii, pressure, temperature, diffusion_creep_parameters);
-              const std::pair<double, double> disl_edot_and_deriv = dislocation_creep.compute_strain_rate_and_derivative(stress_ii, pressure, temperature, dislocation_creep_parameters);
+              const std::pair<double, double> diff_edot_and_deriv = diffusion_creep.compute_strain_rate_and_derivative(stress_ii, pressure, temperature, 1., diffusion_creep_parameters);
```
```suggestion
              const std::pair<double, double> diff_edot_and_deriv = diffusion_creep.compute_strain_rate_and_derivative(stress_ii, pressure, temperature, water_concentration, diffusion_creep_parameters);
```

Would this work since `1.` is the default value for the parameter?

naliboff replied:
Same for places where `1.` is defined instead of `water_concentration`.

naliboff commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -68,17 +69,20 @@ namespace aspect
 
       template <int dim>
-      double
+      double // Add position to args of this function
```
Does this function still need the position argument for this PR? If yes, can you add in the next commit? If not, remove comment?

naliboff commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -91,11 +95,19 @@ namespace aspect
-                                               pressure*p.activation_volume)/
-                                              (constants::gas_constant*temperature)) *
-                                     std::pow(grain_size, p.grain_size_exponent);
+                                           std::exp((p.activation_energy +
```
Was this indentation manual, or done via astyle?

naliboff commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -176,6 +198,17 @@ namespace aspect
+			   "Whether or not to include water fugacity in the diffusion creep rheology. Setting this to false "
+			   "gives the default diffusion creep rheology. If set to true, a compositional field named 'water' "
+			   "must exist, and the material model will query the water content at each mesh point and include "
+			   "that in the calculation for the viscosty. Wet rheological parameters must also be specified.");
```
```suggestion
			   "that in the calculation for the viscosity. Flow law parameters modified for the effects "
			   "hydration state must be specified in the parameter file.");
```

naliboff commented on source/material_model/rheology/diffusion_creep.cc:
```diff
@@ -176,6 +198,17 @@ namespace aspect
+	prm.declare_entry ("Use water fugacity in diffusion creep", "false",
+			   Patterns::Bool(),
+			   "Whether or not to include water fugacity in the diffusion creep rheology. Setting this to false "
+			   "gives the default diffusion creep rheology. If set to true, a compositional field named 'water' "
```
```suggestion
			   "gives the default diffusion creep rheology. If set to true, a compositional field named 'water_concentration' "
```

naliboff commented on source/material_model/rheology/dislocation_creep.cc:
```diff
@@ -90,10 +95,17 @@ namespace aspect
-                                       std::exp((p.activation_energy + pressure*p.activation_volume)/
-                                                (constants::gas_constant*temperature*p.stress_exponent)) *
-                                       std::pow(strain_rate,((1. - p.stress_exponent)/p.stress_exponent));
+                                      std::exp((p.activation_energy + pressure*p.activation_volume)/
```
Indentation update from astyle?

naliboff commented on source/material_model/rheology/dislocation_creep.cc:
```diff
@@ -169,6 +194,17 @@ namespace aspect
+        prm.declare_entry ("Use water fugacity in dislocation creep", "false",
+                           Patterns::Bool(),
+                           "Whether or not to include water fugacity in the dislocation creep rheology. Setting this to false "
+                           "gives the default dislocation creep rheology. If set to true, a compositional field named 'water' "
```
```suggestion
                           "gives the default dislocation creep rheology. If set to true, a compositional field named 'water_concentration' "
```

naliboff commented on source/material_model/rheology/dislocation_creep.cc:
```diff
@@ -169,6 +194,17 @@ namespace aspect
+                           Patterns::Bool(),
+                           "Whether or not to include water fugacity in the dislocation creep rheology. Setting this to false "
+                           "gives the default dislocation creep rheology. If set to true, a compositional field named 'water' "
+                           "must exist, and the material model will query the water content at each mesh point and include "
```
```suggestion
                           "must exist, which is used to modify the dislocation creep viscosity."
```

naliboff commented on tests/visco_plastic_diffusion_dislocation_composite_water_fugacity.prm:
```diff
@@ -0,0 +1,137 @@
+# A test for composite creep in the visco_plastic material model which includes effects from
```
```suggestion
# A test for composite creep in the visco_plastic material model,  which includes the effects of
```

naliboff commented on tests/visco_plastic_diffusion_dislocation_composite_water_fugacity.prm:
```diff
@@ -0,0 +1,137 @@
+# dislocation viscosity multiplied by the concentration of water given as H ppm Si.
+# In this model, half of the box is hydrated with a water concentratin of 100 H ppm Si,
+# and the other half of the box has 0 H ppm Si. The viscosity is therefore reduced in
+# the region where there is water relative to the region that has no water.
```
Can you include a section of comments here (or in the Material Model section), on how the flow law parameters were modified to account from water concentration.

naliboff replied:
Can you also include in the comments the calculation of what water viscosity should be expected.

naliboff commented on tests/visco_plastic_diffusion_dislocation_composite_water_fugacity.prm:
```diff
@@ -0,0 +1,137 @@
+  subsection Visualization
+    set Interpolate output = false
+    set List of output variables = viscosity, strain rate
+    set Output format            = gnuplot
```
```suggestion
    set Output format            = gnuplot
    set Number of grouped files   = 1
```

naliboff reviewed (changes requested):
@danieldouglas92 - thanks for the contribution and this looks to be in pretty good shape. Only few minor comments here and there, largely about formatting and documentation.

bobmyhill commented:
@danieldouglas92 did you want to take another look at this, or shall we close it?

danieldouglas92 commented:
This functionality has been merged in #5678.
